File: bot.py
import os
import discord
import random
from discord.ext import commands
from dotenv import load_dotenv
from groq import Groq

load_dotenv()

DISCORD_TOKEN = os.getenv("DISCORD_TOKEN")
GROQ_API_KEY = os.getenv("GROQ_API_KEY")

client = Groq(api_key=GROQ_API_KEY)

MODEL = "llama-3.3-70b-versatile"

intents = discord.Intents.default()
intents.message_content = True

bot = commands.Bot(command_prefix="!", intents=intents)

# 🧠 채널 기억 저장
memory = {}

# ⚡ 반응 확률 (0~1)
REPLY_CHANCE = 0.35


def ask_ai(channel_id, user_msg):
    if channel_id not in memory:
        memory[channel_id] = []

    history = memory[channel_id]

    messages = [
        {
            "role": "system",
            "content": "너는 친구처럼 말하는 AI야. 반말, 짧고 자연스럽게 말해. 너무 AI처럼 말하지 마."
        }
    ]

    messages += history[-12:]
    messages.append({"role": "user", "content": user_msg})

    response = client.chat.completions.create(
        model=MODEL,
        messages=messages,
        temperature=0.8
    )

    reply = response.choices[0].message.content

    # 기억 저장
    history.append({"role": "user", "content": user_msg})
    history.append({"role": "assistant", "content": reply})

    memory[channel_id] = history

    return reply


@bot.event
async def on_ready():
    print(f"🔥 봇 실행됨: {bot.user}")


@bot.event
async def on_message(message):
    if message.author.bot:
        return

    content = message.content.strip()

    # 🍜 1. "떡볶이" 있으면 무조건 반응
    force_reply = False

    if content.startswith("떡볶이"):
        force_reply = True
        content = content.replace("떡볶이", "", 1).strip()

    # 🍜 2. 없으면 랜덤 확률로 반응
    if not force_reply:
        if random.random() > REPLY_CHANCE:
            return

    # 아무 내용 없으면 무시
    if not content:
        return

    async with message.channel.typing():
        reply = ask_ai(message.channel.id, content)

        if len(reply) > 2000:
            reply = reply[:1990] + "..."

        await message.reply(reply)

    await bot.process_commands(message)


bot.run(DISCORD_TOKEN)

File: bot.js
require("dotenv").config();

const { Client, GatewayIntentBits } = require("discord.js");
const { GoogleGenAI } = require("@google/genai");
const fs = require("fs");

// ===== AI =====
const ai = new GoogleGenAI({
  apiKey: process.env.GEMINI_API_KEY,
});

// ===== Discord =====
const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
  ],
});

// ===== 관리자 =====
const ADMIN_ID = process.env.ADMIN_ID;

// ===== memory =====
const FILE = "./memory/memory.json";

function load() {
  if (!fs.existsSync(FILE)) return {};
  return JSON.parse(fs.readFileSync(FILE, "utf8"));
}

function save(data) {
  fs.writeFileSync(FILE, JSON.stringify(data, null, 2));
}

let memory = load();

// ===== 말투 시스템 =====
const styles = {
  기본: "너는 자연스럽고 친절한 AI다.",

  불닭맛:
    "너는 매우 텐션 높고 매운 느낌으로 말한다. 과장된 표현을 사용하지만 욕설은 하지 않는다.",

  차분: "너는 매우 차분하고 짧고 안정적으로 말한다.",

  존댓말: "항상 정중한 존댓말로만 말한다.",

  광기: "약간 미친 듯한 텐션이지만 위험하지 않은 재미있는 말투로 답한다."
};

// ===== AI 호출 =====
async function askAI(id, text) {
  const user = memory[id] || { history: [], style: "기본" };

  const style = styles[user.style] || styles["기본"];

  const result = await ai.models.generateContent({
    model: "gemini-2.5-flash",
    contents: [
      {
        role: "user",
        parts: [
          {
            text: `
${style}

이 규칙을 반드시 유지해라.

대화 기록:
${JSON.stringify(user.history.slice(-8))}

유저: ${text}
            `
          }
        ]
      }
    ],
  });

  const reply =
    result.candidates?.[0]?.content?.parts?.[0]?.text || "오류 발생";

  memory[id] = {
    style: user.style || "기본",
    history: [
      ...user.history,
      { role: "user", parts: [{ text }] },
      { role: "model", parts: [{ text: reply }] }
    ].slice(-30)
  };

  save(memory);

  return reply;
}

// ===== 메시지 분할 =====
function split(text) {
  const arr = [];
  for (let i = 0; i < text.length; i += 1900) {
    arr.push(text.slice(i, i + 1900));
  }
  return arr;
}

// ===== ready =====
client.once("ready", () => {
  console.log(`로그인됨: ${client.user.tag}`);
});

// ===== message =====
client.on("messageCreate", async (message) => {
  if (message.author.bot) return;

  const id = message.author.id;
  const content = message.content;

  const isAdmin = id === ADMIN_ID;

  // =========================
  // 📌 관리자 기능
  // =========================
  if (content === "!shutdown") {
    if (!isAdmin) return message.reply("권한 없음");
    await message.reply("종료");
    process.exit(0);
  }

  if (content === "!resetall") {
    if (!isAdmin) return message.reply("권한 없음");
    memory = {};
    save(memory);
    return message.reply("전체 초기화 완료");
  }

  if (content.startsWith("!say ")) {
    if (!isAdmin) return;
    return message.channel.send(content.replace("!say ", ""));
  }

  // =========================
  // 📌 말투 변경
  // =========================
  if (content.startsWith("!말투 ")) {
    const style = content.replace("!말투 ", "");

    if (!styles[style]) {
      return message.reply(
        "가능: 기본 / 불닭맛 / 차분 / 존댓말 / 광기"
      );
    }

    memory[id] = memory[id] || { history: [], style: "기본" };
    memory[id].style = style;
    save(memory);

    return message.reply(`말투 변경됨: ${style}`);
  }

  // =========================
  // 📌 일반 명령어
  // =========================
  if (content === "!help") {
    return message.reply("!말투, !reset, !shutdown(관리자)");
  }

  if (content === "!reset") {
    memory[id] = { history: [], style: "기본" };
    save(memory);
    return message.reply("초기화 완료");
  }

  // =========================
  // 📌 AI 응답
  // =========================
  try {
    const reply = await askAI(id, content);

    for (const part of split(reply)) {
      await message.reply(part);
    }
  } catch (e) {
    console.error(e);
    message.reply("AI 오류");
  }
});

client.login(process.env.DISCORD_TOKEN);
